Add possibility of resuming multinode jobs

zv0n opened the pull request:
#### Describe Bug or Feature
I have added possibility of resuming jobs that are running on multiple nodes. It is already possible to resume a job running on single node after mom restarts, so allowing the same for multinode jobs seems like a good idea. This is done when `-p` flag is passed on command line.

Possible problems with the implementation:
  - If you run `pbsdsh` while one of the nodes is down it enters an infinite loop as it doesn't check for node availability
  - If you restart mother superior while `pbsdsh` is running, it will exit with error as it loses connection with mother superior
  - Starting `pbsdsh` right after a node restart may result in error on task spawn because the task can be sent to the node before the node rejoins the job (and receives ports through which it should communicate)


#### Describe Your Change
  - Mother superior now saves and recovers `ji_ports`, so it can communicate with demux program after recovering the job.
  - When a non-superior mom recovers multinode job, it sends `IM_RECONNECT_TO_MS` to mother superior, MS then sends `IM_JOIN_RECOV_JOB` to all sisters along with port numbers through which they should communicate
  - When MS recovers multinode job, it sends `IM_JOIN_RECOV_JOB` along with port numbers to all sisters
  - Obit information (if there is any) is saved/recover along with tasks, so moms know wheter to send an obit after task completion
  - Multinode jobs now check whether they have a connection to MS inside `scan_for_exiting`, so when they send an obit they can be sure that it is delivered

#### Attach Test and Valgrind Logs/Output
[test_multinode_restart.txt](https://github.com/openpbs/openpbs/files/5061012/test_multinode_restart.txt)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/openpbs/openpbs?pullRequest=1955) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/openpbs/openpbs?pullRequest=1955) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/openpbs/openpbs?pullRequest=1955) it.</sub>

zv0n commented:
Hi,
Can someone please remove the random redundant Travis job (that doesn't seem to be starting at all)?
Thanks!

arungrover commented:
Hi @zv0n, I think I have seen this problem with Travis in the past. What might work is that either you can close the PR and reopen it OR force push your branch. In both these cases, Travis will run again but I think it will get rid of that redundant job.

zv0n commented:
Thanks @arungrover, seems like force pushing fixed it!

zv0n commented:
Can someone take a look at this change, please?

bayucan commented on src/include/job.h:
```diff
@@ -501,6 +501,8 @@ struct job {
 	int		ji_ports[2];	/* ports for stdout/err */
 	enum	bg_hook_request	ji_hook_running_bg_on; /* set when hook starts in the background*/
+	int		ji_msconnected; /* 0 - not connected, 1 - connected */
+	pbs_list_head	ji_multinodejobs;	/* links to recovered multinode jobs */
```
Please initialize ji_msconnected in server/job_func.c:job_alloc() under #ifdef PBS_MOM section.
Initialize also ji_multinodejobs in server/job_func.c:job_alloc() by calling CLEAR_HEAD(ji_multinodejobs).

bayucan commented on test/tests/functional/pbs_node_jobs_restart_multinode.py:
```diff
@@ -0,0 +1,118 @@
+
+        # Make sure moms are running with -p flag
+        momA.restart(args=['-p'])
+        momB.restart(args=['-p'])
```
What happens if -p is not specified by at least one of the moms? Please add a test case for this.

bayucan commented on test/tests/functional/pbs_node_jobs_restart_multinode.py:
```diff
@@ -0,0 +1,118 @@
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid)
+
+        # only restart 1 node as restarting momA would result in a pbsdsh crash
+        momB.restart(args=['-p'])
```
Please add a test case where momB is prematurely killed, as in mom.signal("-KILL"), and then issue a mom.restart("-p"). Also a case where primary mom is also killed prematurely, and then restart with -p. Is the behavior as you expected?

hirenvadalia commented on src/resmom/mom_comm.c:
```diff
@@ -217,6 +219,41 @@ pbs_jobnodevoid_t	job_free_node = NULL;
+	int i = 0;
+	while ((i = write(fd, write_ptr, size)) != size) {
+		if ((i < 0) && (errno == EINTR)) {	/* retry the write */
+			if (lseek(fd, (off_t)0, SEEK_SET) < 0) {
```
Doing lseek 0 here can lead to corrupted data in .TK file...
For example, if ensure_write is called for `obit->oe_u.oe_tm.oe_node` (4th write in file), and if we do lseek 0 then will loose data for `ptask->ti_qs`, `pobit->oe_type`, `pobit->oe_u.oe_tm.oe_fd`

hirenvadalia commented on src/resmom/mom_comm.c:
```diff
@@ -278,8 +315,26 @@ task_save(pbs_task *ptask)
 			return (-1);
 		}
 	}
+	obitent *pobit = (obitent *)GET_NEXT(ptask->ti_obits);
```
I am not sure but ptask->ti_obits is list, and we are only storing first? don't we need to store others in list?

zv0n replied:
You're probably right, I thought that obits are only created when MS first sends a task, but now I see they can be created in other ways as well. I'll rethink how the obits are saved, thanks for the catch!

zv0n commented:
> @zv0n : Do you also have an accompanying design for this? It would be better to add a design under https://openpbs.atlassian.net/wiki/spaces/PD/pages outlining your approach of when pbs_mom is started with -p option, multinode jobs can also be recovered. Then announce it to the community form: https://community.openpbs.org/ for comments. People in the community can chime in and suggest improvements to the design...

@bayucan I don't have a design, sorry. I did this because we needed this functionality at my workplace, so I started coding before I looked at how merging to the project works. I'll go ahead and create one now.

zv0n commented on src/include/job.h:
```diff
@@ -501,6 +501,8 @@ struct job {
 	int		ji_ports[2];	/* ports for stdout/err */
 	enum	bg_hook_request	ji_hook_running_bg_on; /* set when hook starts in the background*/
+	int		ji_msconnected; /* 0 - not connected, 1 - connected */
+	pbs_list_head	ji_multinodejobs;	/* links to recovered multinode jobs */
```
I believe I'm already initializing these variables in job_func.c:355,356, should I do something more than that?

Since ji_multinodejobs only contains pointers to jobs that are already managed, will CLEAR_HEAD(ji_multinodejobs) suffice in job_free()?

bayucan replied:
I missed that you were already initializing them in job_func.c under job_alloc() so that's good. For the job_free(), yes, just do the CLEAR_HEAD.

zv0n commented on src/include/job.h:
```diff
@@ -612,6 +614,8 @@ struct job {
 			tm_host_id ji_nodeidx;	/* my node id */
 			tm_task_id ji_taskidx;	/* generate task id's for job */
+			int ji_stdout;
+			int ji_stderr;
```
No other vars in ji_extended are allocated in job_alloc(), should I really do that?

zv0n commented on src/resmom/mom_comm.c:
```diff
@@ -3186,6 +3289,9 @@ im_request(int stream, int version)
 
 			/* np is set from job_nodes_inner */
 
+			if (check_ms(stream, pjob))
```
Sorry, I forgot to revert this change (I did it during development when connection to MS was not working and I was trying various fixes). I'll revert it, but I think the early check isn't really needed, as it was running check_ms(stream, NULL) and check_ms always returns FALSE when pjob == NULL, so it seems like useless line to me

hirenvadalia commented on src/resmom/mom_comm.c:
```diff
@@ -278,8 +317,43 @@ task_save(pbs_task *ptask)
+		obitcount++;
+		if (ensure_write(fds, (char *)&pobit->oe_type, sizeof(pobit->oe_type)) != 0)
+			goto failedwrite;
+		if (ensure_write(fds, (char *)&pobit->oe_u.oe_tm.oe_fd, sizeof(pobit->oe_u.oe_tm.oe_fd)) != 0)
```
@bayucan Please correct me if I am wrong...
but @zv0n I don't think saving obits help us regain connection with clients waiting for obits as if mom process goes then oe_fd will be released and clients (aka pbsdsh or pbs_tmrsh) connected to oe_fd will get disconnect error right? like below:
```
pbsroot@node1:~$ qsub -I
qsub: waiting for job 2.node1 to start
qsub: job 2.node1 ready

pbsroot@node1:~$ pbsdsh sleep 100 
pbsdsh: Event poll failed, error not connected (17002)
```
I just killed pbs_mom when sleep 100 process started

zv0n replied:
Yes, that's one of the issues I've written about in the initial pull request comment, pbsdsh fails when mother superior is restarted. But if you restart one of the sisters, pbsdsh continues going and expects obit from the sisters, so if we don't save the obit information, it would result in an infinite loop within pbsdsh.

In this commit I wanted to create server-side (mom) support for the multi-node resume and thought client side might be done in a different commit.

This problem is within pbsdsh only and I think it should be fixable, for example mpirun (at least during my testing) didn't error out during MS restart and managed to reconnect and after receiving obits terminate successfuly.

hirenvadalia replied:
So if I understand correctly @zv0n, then if we fix clients not to error out and wait for MS or sister nodes to come up and reconnect, then also we don't need save obit info, because once clients reconnects it will do tm_obit (to register obit) and tm_poll (wait for obit to finish) again right?

zv0n replied:
@hirenvadalia Yes, that sounds right. If the clients can recognize when a node reconnects, so they can delete the old event and register a new obit, that would remove the need to save obits.

hirenvadalia replied:
exactly @zv0n!, only thing is clients has to try (may be forever or for some timeout) reconnecting to node... and once connection re-establish, no need to delete old event, just register obit for given task and wait for its finish

hirenvadalia replied:
Also doing above will eliminate all possible problems which you mentioned in PR description...

zv0n replied:
Wouldn't the old event need to be deleted? `tm_obit` calls `add_event` inside the function body, which allocates a new event, wouldn't that be a problem? If we don't want to delete the old event I think we might need to add a new function that only re-sends the obit number without creating a new event.
I've only skimmed through the function, so I might be wrong about how it works though.

hirenvadalia replied:
Since obits are not saved in .TK file so on mom restart mom wouldn't have that information, if we save obits in .TK and recover that on mom restart, then only we need to delete that, but as we don't need to save obits so no need of deleting it... HTH!

zv0n replied:
I see! Great, thanks!

zv0n replied:
So, @hirenvadalia, do you think I should remove the obit-saving parts and leave this PR as is? Because so far I have no idea how to make pbsdsh operate without an open socket to MS 😅

hirenvadalia replied:
@zv0n, sorry for delay, I think we can proceed with this PR after removing obit-saving parts but without fixing pbsdsh/pbs_tmrsh, will not get full resume of multinode.
Based on your one of above comment, where you say mpirun was able to reconnect on MS restart and continue, so if we don't fix pbsdsh and pbs_tmrsh then only multinode jobs which uses mpirun part will able resume i think.

zv0n replied:
@hirenvadalia I've removed the obit saving bits and did a test run, as you expected mpirun works without a problem, pbsdsh gets stuck on waiting for obit.

hirenvadalia replied:
Thanks @zv0n... will continue code and design review...

bayucan commented:
Thanks @zv0n for creating the design.

subhasisb commented:
hi @zv0n I was looking at your proposal. I looked at your design page, but could not go to the associated design discussion page - it shows "oops topic not found". Hence adding my thoughts here;

I was trying to understand your primary usecase to make this change:
Are you trying to handle the case where only pbs_mom is restarted/upgraded? Or for a host level restart. If it is a host level restart/OS-update etc that needs a reboot, pbsdsh, or MPI job will typically die anyway, so allowing mom to restart with -p to preserve jobs is not useful there. 
PBS supports ways to make node maintenance easy, by allowing to turn nodes offline to drain off jobs from them, or to enable admin-suspend: maintenance_jobs in the guide: https://www.altair.com/pdfs/pbsworks/PBS19.2.3_BigBook.pdf

Also, there is a thread about new maintenance windows enhancement: 
https://openpbs.atlassian.net/wiki/spaces/PD/pages/879493121/Node+maintenance+window+enhancement

Please look through these and see if they satisfy your use cases already or not?

I do realize that your effort is to allow restarting of mom while preserving multinode jobs. The part i am trying to understand is that "is it really necessary, when there are ways to take a node down for maintenance in an orderly fashion anyway?"

vchlum commented:
Hi @subhasisb , my colleague @zv0n works on this because we would appreciate this feature/fix for the upgrading process, and I think this could help others too.

Of course, this can not cover the host-level restart need. In such a case, the maintenance windows are suitable.

Our current solution: When we need to update the PBS service on nodes (because applying some patch or bug fix), we often do the "live update". Right now, we have the following patch in use: After upgrading the PBS package on the system, a signal is sent to mom. The mom will wait until no more multi-node jobs are running on the node. Once no multi-node job is present the self restart is engaged (with -p). This has obviously a huge drawback - the restart may never happen because a new multi-node job could start before the last one is finished. Not ideal.

I think it would be good and useful for all to be able to update (minor? versions of) PBS without setting any maintenance. We could just install a new version of the PBS package and the post-install could initiate the restart and that's it. This "live update" would be much simpler for admins. It would also increase the utilization of nodes because you do not need to drain nodes before updating. So, no resources are wasted during the PBS package update process.

subhasisb commented:
> Our current solution: When we need to update the PBS service on nodes (because applying some patch or bug fix), we often do the "live update". Right now, we have the following patch in use: After upgrading the PBS package on the system, a signal is sent to mom. The mom will wait until no more multi-node jobs are running on the node. Once no multi-node job is present the self restart is engaged (with -p). This has obviously a huge drawback - the restart may never happen because a new multi-node job could start before the last one is finished. Not ideal.

Actually, if you set the node to "offline" no new jobs will be sent to the mom. So eventually, the node will be free. I do understand  that even that could take a good amount of time depending on how long the currently running job takes to complete.

> I think it would be good and useful for all to be able to update (minor? versions of) PBS without setting any maintenance. We could just install a new version of the PBS package and the post-install could initiate the restart and that's it. This "live update" would be much simpler for admins. It would also increase the utilization of nodes because you do not need to drain nodes before updating. So, no resources are wasted during the PBS package update process.

yes, this, of course, makes sense. I was just thinking whether lots of sites really need this. But, i understand if you do.

bhroam commented:
@vchlum another one of our customers handles rolling updates with provisioning.  They even handle OS updates this way.  They set the default aoe of the new jobs to the new node image.  Eventually a new job will run with the new aoe and that will trigger the node to be provisioned.  This will work better for them than for you because they allocate their nodes exclusively.  In your case, the scheduler will only run jobs with shared aoe values on the same nodes.  Eventually all of the older aoe jobs will be finished and a new aoe will run.  The node will eventually provision.

vchlum commented:
Thx @bhroam for sharing the solution. A solution such as this is not suitable for our site. We have heterogeneous nodes and heterogeneous jobs. We allow up to one-month walltime.

Since the -p parameter already works for singlenode job, I also think this is a logical improvement.

zv0n commented:
My forum post got approved, so we can move future discussion there if need be - https://community.openpbs.org/t/option-to-resume-multinode-jobs/2244/2

zv0n commented:
Forgot about that, sorry, it's removed now

hirenvadalia reviewed (approved):
@zv0n I don't have any further comments, but you still have few open question/comments from @bayucan. Also you need to update design page as suggestion by @bayucan.

zv0n commented:
Hi, sorry for the long delay, I was on vacation.
@bayucan I think I've fixed all problems with code you've found and just added your requested tests. I've removed the test which restarted during pbsdsh since we've decided to remove obit saving and that test would either fail or enter an infinite loop

bayucan commented:
@zv0n : It looks good now. Travis check seemed to be stuck. Please rebase your branch to the latest master.

zv0n commented:
@bayucan I've rebased to master, hopefully Travis check will be happy and cooperative this time around :D

hirenvadalia commented on src/resmom/start_exec.c:
```diff
@@ -5927,6 +5927,52 @@ job_nodes(struct job *pjob)
+
+		int stream = np->hn_stream;
+		im_compose(stream, pjob->ji_qs.ji_jobid,
+			pjob->ji_wattr[(int)JOB_ATR_Cookie].at_val.at_str,
```
@zv0n sorry but one very small change: (as now things have changed in mainline about how we access job attributes using getters/setters):
Can you please repalce this line with `get_jattr_str(pjob, JOB_ATR_Cookie)`?

hirenvadalia reviewed (approved):
Thanks @zv0n. LGTM.

bayucan reviewed (approved):
Code looks good.
